rust/tools/authorization-logic/raksha_examples/VisionPipelineAuthLogic

```diff
@@ -0,0 +1,153 @@
+    // name of a product has taken place, but only when the user has selected 
+    // an object.
+    //
+    // It might look like this ideally:
+    //
+    // "ThirdPartyPrivacyReviewer" canSay becomes(pathX, "raw_video_tag",
+    //              "product_id_tag") :-
+    //     objectSelected("EndUser", pathX),
+    //     isAccessPath(pathX).
+    //
+    // For now it has to be done like this:
+    "ThirdPartyPrivacyReviewer" canSay declassifies(pathX, "raw_video_tag") :-
+        hasTag(pathX, "product_id_tag"),   // raw_video_tag is only removed if 
+        objectSelected("EndUser", pathX),  // product_id_tag is added
+        isAccessPath(pathX).
+
+    "ThirdPartyPrivacyReviewer" canSay
+        objectSelected("EndUser", "identified_product").
+
+    // Delegates TPPR the ability to apply tags anywhere
+    "ThirdPartyPrivacyReviewer" canSay hasTag(x, y) :- 
+        isAccessPath(x), isTag(y).
+
+    // Delegate the right to claim that particles are "Local" to TPPR.
+    // (We can think of the principal "LocalDevice" as a group)
+    "ThirdPartyPrivacyReviewer" canSay principalX canActAs "LocalDevice" :-
+        isPrincipal(principalX).
+
+    "LocalDevice" may("do_pure_computation", "raw_video_tag").
+    "LocalDevice" may("egress_to_shopping_service", "product_id_tag").
+
+    // Could also give different permissions to RemoteService for once it
+    // has "product_id_tag".
+
+    // These are rules we should get rid of in the real thing.
+    // The universe relations probably don't need to be decentralized. We could
+    // have rules that populate a "centralized" universe and passes this 
+    // to all the principals. (Or there could be other ways to deal with this).
+    isPrincipal("VideoDataSource").
+    isPrincipal("ImageDetector").
+    isPrincipal("ImageSelector").
+    isPrincipal("ProductIdOutput").
+}
+
+"ThirdPartyPrivacyReviewer" says {
+    "VideoDataSource" canActAs "LocalDevice".
+    "ImageDetector" canActAs "LocalDevice".
+    "ImageSelector" canActAs "LocalDevice".
+    "ProductIdOutput" canActAs "LocalDevice".
+    
+    // TPPR delegates the data transformation to ImageSelector
+    // "ImageSelector" canSay becomes(pathX, "raw_video_tag", "product_id_tag") :-
+    //     isAccessPath(pathX).
+    "ImageSelector" canSay declassifies(pathX, "raw_video_tag") :-
+        hasTag(pathX, "product_id_tag"),   // raw_video_tag is only removed if 
+        objectSelected("EndUser", pathX),  // product_id_tag is added
+        isAccessPath(pathX).
+    "ImageSelector" canSay hasTag(pathX, "product_id_tag") :- 
+        isAccessPath(pathX).
+    "ImageSelector" canSay objectSelected("EndUser", pathX) :-
+        isAccessPath(pathX).
+}
+
+// Maybe we want this part of the policy to sit closer to the particles
+// because sometimes relations will refer to path names ?
+// (We can ostensibly do this as long as we can write
+// `claim <RelationName(...)>` in manifests)
+"VideoDataSource" says "VideoDataSource"
```

Comment:
Why is "VideoDataSource" repeated?

Reply:
`"VideoDataSource" will("do_pure_computation", "raw_video_data_tag")` is very similar to `will("VideoDataSource", "do_pure_computation", "raw_video_tag") except that when it's written this way it can interact with the `canActAs` construct to pass the `will(...)` predicate to other principals (like the "LocalDevice" group). Predicates with the syntax `<Principal>    <Relation>(<args>)` (and with this interaction with `canActAs`) are kind of like "properties" of the principal and `canActAs` is kind of like saying one principal can pick up the properties of another.

So that's the inner instance of "VideoDataSource".

The outer "VideoDataSource"  before `says` is because VDS is the principal claiming that VDS has this property. VDS is describing its own behavior in this example.